manager/orchestrator/globaljob/reconciler.go

```diff
@@ -0,0 +1,176 @@
+package globaljob
+
+import (
+	"fmt"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/manager/constraint"
+	"github.com/docker/swarmkit/manager/orchestrator"
+	"github.com/docker/swarmkit/manager/state/store"
+	gogotypes "github.com/gogo/protobuf/types"
+)
+
+type reconciler interface {
+	ReconcileService(id string) error
+}
+
+type reconcilerObj struct {
+	store *store.MemoryStore
+}
+
+func newReconciler(store *store.MemoryStore) reconciler {
+	return &reconcilerObj{
+		store: store,
+	}
+}
+
+func (r *reconcilerObj) ReconcileService(id string) error {
+	var (
+		service *api.Service
+		cluster *api.Cluster
+		tasks   []*api.Task
+		nodes   []*api.Node
+		viewErr error
+	)
+
+	// we need to first get the latest iteration of the service, its tasks, and
+	// the nodes in the cluster.
+	r.store.View(func(tx store.ReadTx) {
+		service = store.GetService(tx, id)
+		if service == nil {
+			return
+		}
+
+		// getting tasks with FindTasks should only return an error if we've
+		// made a mistake coding; there's no user-input or even reasonable
+		// system state that can cause it. If it returns an error, we'll just
+		// panic and crash.
+		tasks, viewErr = store.FindTasks(tx, store.ByServiceID(id))
+		if viewErr != nil {
+			return
+		}
+
+		// same as with FindTasks
+		nodes, viewErr = store.FindNodes(tx, store.All)
+		if viewErr != nil {
+			return
+		}
+
+		clusters, _ := store.FindClusters(tx, store.All)
```

Comment:
Previous PR asked about a helper function. IMHO, not worth rooting around in the store package and changing even small things. It's just better to leave well enough alone.